sanic/exceptions.py

```diff
@@ -9,103 +10,292 @@ class RequestCancelled(CancelledError):
 
 
 class ServerKilled(Exception):
-    ...
+    """
+    Exception Sanic server uses when killing a server process for something
+    unexpected happening.
+    """
 
 
 class SanicException(Exception):
+    """
+    Generic exception that will generate an HTTP response when raised
+    in the context of a request lifecycle.
+
+    Usually it is best practice to use one of the more specific exceptions
+    than this generic. Even when trying to raise a 500, it is generally
+    preferrable to use :class:`.ServerError`
+
+    .. code-block:: python
+
+        raise SanicException(
+            "Something went wrong",
+            status_code=999,
+            context={
+                "info": "Some additional details",
+            },
+            headers={
+                "X-Foo": "bar"
+            }
+        )
+
+    :param message: The message to be sent to the client. If ``None``
+        then the appropriate HTTP response status message will be used
+        instead, defaults to None
+    :type message: Optional[Union[str, bytes]], optional
+    :param status_code: The HTTP response code to send, if applicable. If
+        ``None``, then it will be 500, defaults to None
+    :type status_code: Optional[int], optional
+    :param quiet: When ``True``, the error traceback will be suppressed
+        from the logs, defaults to None
+    :type quiet: Optional[bool], optional
+    :param context: Additional mapping of key/value data that will be
+        sent to the client upon exception, defaults to None
+    :type context: Optional[Dict[str, Any]], optional
+    :param extra: Additional mapping of key/value data that will NOT be
+        sent to the client when in PRODUCTION mode, defaults to None
+    :type extra: Optional[Dict[str, Any]], optional
+    :param headers: Additional headers that should be sent with the HTTP
+        response, defaults to None
+    :type headers: Optional[Dict[str, Any]], optional
+    """
+
+    status_code: int
+    quiet: Optional[bool]
+    headers: Dict[str, str]
     message: str = ""
 
     def __init__(
         self,
         message: Optional[Union[str, bytes]] = None,
         status_code: Optional[int] = None,
+        *,
```

Comment:
Yes, this is a breaking change that I think we should be making. It was a mistake that lead to a messy API allowing these to not be keyword-required.